File: 2018/06-nearest-neighbour-coordinates/python/star2.py
#!/usr/bin/env python3

from collections import defaultdict


def distance(c1, c2):
    return abs(c1[0] - c2[0]) + abs(c1[1] - c2[1])


with open('input') as f:
    lines = f.read().splitlines()

coords = [tuple([int(i) for i in l.split(', ')]) for l in lines]

xs = [c[0] for c in coords]
ys = [c[1] for c in coords]

top_left = (min(xs), min(ys))
bottom_right = (max(xs), max(ys))

def inside_edge(coord, top_left, bottom_right):
    x, y = coord
    x_inside = top_left[0] < x and x < bottom_right[0]
    y_inside = top_left[1] < y and y < bottom_right[1]
    return x_inside and y_inside

finite_coords = [c for c in coords if inside_edge(c, top_left, bottom_right)]

reference_distance = 10000
region_size = 0
for i in range(top_left[0]+1, bottom_right[0]):
    for j in range(top_left[1]+1, bottom_right[1]):
        total_dist = 0
        for coord in coords:
            dist = distance((i,j), coord)
            total_dist += dist
        if total_dist < reference_distance:
            region_size += 1

print(region_size)

File: 2018/06-nearest-neighbour-coordinates/rust/star2.rs
use std::cmp::min;
use std::cmp::max;

const INPUT: &str = include_str!("../input");

fn distance(point1: (i32, i32), point2: (i32, i32)) -> i32 {
    return (point1.0 - point2.0).abs() + (point1.1 - point2.1).abs()
}

fn main() {
    let mut top_left = (i32::max_value(), i32::max_value());
    let mut bottom_right = (i32::min_value(), i32::min_value());
    let mut coordinates: Vec<(i32, i32)> = Vec::new();
    for line in INPUT.lines() {
        let ints: Vec<i32> = line.split(", ")
            .map(|s| s.parse()
            .unwrap())
            .collect();

        top_left = (
            min(top_left.0, ints[0]),
            min(top_left.1, ints[1])
        );
        bottom_right = (
            max(bottom_right.0, ints[0]),
            max(bottom_right.1, ints[1])
        );
        coordinates.push((ints[0], ints[1]));
    }

    const MAX_DISTANCE: i32 = 10000;
    let mut region_size = 0;
    for x in (top_left.0)..(bottom_right.0) {
        for y in (top_left.1)..(bottom_right.1) {
            let mut total_distance = 0;
            for coordinate in coordinates.as_slice() {
                total_distance += distance(*coordinate, (x, y));
            }
            if total_distance < MAX_DISTANCE {
                region_size += 1;
            }
        }
    }

    println!("{}", region_size);
}
